// @sanity/ui/dist/cjs/theme/studioTheme/helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.screen = exports.multiply = void 0;
var color_blend_1 = require("color-blend");
var parse_color_1 = __importDefault(require("parse-color"));
function multiply(bg, fg) {
    var _a = parse_color_1.default(bg).rgb;
    var _b = parse_color_1.default(fg).rgb;
    var _c = color_blend_1.multiply({
        r: _a[0],
        g: _a[1],
        b: _a[2],
        a: 1,
    }, {
        r: _b[0],
        g: _b[1],
        b: _b[2],
        a: 1,
    }), r = _c.r, g = _c.g, b = _c.b;
    return "rgb(" + r + "," + g + "," + b + ")";
}
exports.multiply = multiply;
function screen(bg, fg) {
    var _a = parse_color_1.default(bg).rgb;
    var _b = parse_color_1.default(fg).rgb;
    var _c = color_blend_1.screen({
        r: _a[0],
        g: _a[1],
        b: _a[2],
        a: 1,
    }, {
        r: _b[0],
        g: _b[1],
        b: _b[2],
        a: 1,
    }), r = _c.r, g = _c.g, b = _c.b;
    return "rgb(" + r + "," + g + "," + b + ")";
}
exports.screen = screen;
