// @sanity/ui/dist/cjs/atoms/select/styles.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.select = void 0;
var styled_components_1 = require("styled-components");
var styles_1 = require("../../styles");
var root = function () {
    return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    position: relative;\n    width: stretch;\n\n    &&:not([hidden]) {\n      display: inline-block;\n    }\n  "], ["\n    position: relative;\n    width: stretch;\n\n    &&:not([hidden]) {\n      display: inline-block;\n    }\n  "])));
};
var inputBase = function (_a) {
    var theme = _a.theme;
    var font = theme.sanity.fonts.text;
    return styled_components_1.css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    -webkit-font-smoothing: antialiased;\n    appearance: none;\n    border: 0;\n    font-family: ", ";\n    color: inherit;\n    width: 100%;\n    outline: none;\n  "], ["\n    -webkit-font-smoothing: antialiased;\n    appearance: none;\n    border: 0;\n    font-family: ", ";\n    color: inherit;\n    width: 100%;\n    outline: none;\n  "])), font.family);
};
var inputColor = function (_a) {
    var theme = _a.theme;
    var color = theme.sanity.color.input;
    return styled_components_1.css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n    background-color: ", ";\n    color: ", ";\n    box-shadow: inset 0 0 0 1px ", ";\n\n    @media (hover: hover) {\n      &:not(:disabled):hover {\n        background-color: ", ";\n        color: ", ";\n        box-shadow: inset 0 0 0 1px ", ";\n      }\n    }\n\n    &:not(:disabled):focus {\n      box-shadow: 0 0 0 1px var(--card-bg-color), 0 0 0 3px var(--card-focus-ring-color);\n    }\n\n    &:not(:disabled):focus:not(:focus-visible) {\n      box-shadow: none;\n    }\n\n    &:disabled {\n      opacity: 1;\n      background-color: ", ";\n      color: ", ";\n      box-shadow: inset 0 0 0 1px ", ";\n    }\n  "], ["\n    background-color: ", ";\n    color: ", ";\n    box-shadow: inset 0 0 0 1px ", ";\n\n    @media (hover: hover) {\n      &:not(:disabled):hover {\n        background-color: ", ";\n        color: ", ";\n        box-shadow: inset 0 0 0 1px ", ";\n      }\n    }\n\n    &:not(:disabled):focus {\n      box-shadow: 0 0 0 1px var(--card-bg-color), 0 0 0 3px var(--card-focus-ring-color);\n    }\n\n    &:not(:disabled):focus:not(:focus-visible) {\n      box-shadow: none;\n    }\n\n    &:disabled {\n      opacity: 1;\n      background-color: ", ";\n      color: ", ";\n      box-shadow: inset 0 0 0 1px ", ";\n    }\n  "])), color.default.enabled.bg, color.default.enabled.fg, color.default.enabled.border, color.default.hovered.bg, color.default.hovered.fg, color.default.hovered.border, color.default.disabled.bg, color.default.disabled.fg, color.default.disabled.border);
};
var textSize = function (size) {
    return {
        fontSize: styles_1.rem(size.fontSize),
        lineHeight: styles_1.rem(size.lineHeight),
    };
};
var inputTextSize = function (_a) {
    var theme = _a.theme, uiSize = _a.uiSize;
    var sizes = theme.sanity.fonts.text.sizes;
    return styles_1.responsive(theme.sanity.media, styles_1.getResponsiveProp(uiSize).map(function (sizeIndex) { return textSize(sizes[sizeIndex] || sizes[2]); }));
};
var input = function () { return [
    styles_1.responsiveRadiusStyle,
    inputBase,
    inputColor,
    inputTextSize,
    styles_1.responsiveInputPaddingIconRightStyle,
]; };
var iconBox = function (_a) {
    var theme = _a.theme;
    var color = theme.sanity.color.input;
    return styled_components_1.css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n    pointer-events: none;\n    position: absolute;\n    top: 0;\n    right: 0;\n\n    select:hover + & {\n      color: ", ";\n    }\n\n    select:disabled + & {\n      color: ", ";\n    }\n  "], ["\n    pointer-events: none;\n    position: absolute;\n    top: 0;\n    right: 0;\n\n    select:hover + & {\n      color: ", ";\n    }\n\n    select:disabled + & {\n      color: ", ";\n    }\n  "])), color.default.hovered.fg, color.default.disabled.fg);
};
exports.select = {
    root: root,
    input: input,
    iconBox: iconBox,
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
