// @sanity/ui/dist/es/components/dialog/dialog.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { forwardRef, useCallback, useEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components';
import { Box, Button, Card, Container, Flex, Text } from '../../atoms';
import { focusFirstDescendant, focusLastDescendant } from '../../helpers';
import { useClickOutside, useGlobalKeyDown } from '../../hooks';
import { Layer, Portal, useLayer } from '../../utils';
const Root = styled(Layer)(({ theme }) => {
    const color = theme.sanity.color.base;
    return css `
    && {
      position: absolute;
    }

    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    &&:not([hidden]) {
      display: flex;
    }
    align-items: center;
    justify-content: center;
    padding: 1.25em;
    outline: none;
    background: ${color.shadow.penumbra};
  `;
});
const DialogContainer = styled(Container) `
  width: 100%;
  height: 100%;
  &&:not([hidden]) {
    display: flex;
  }
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
const DialogCardRoot = styled(Card) `
  width: 100%;
  min-height: 0;
  max-height: 100%;
  overflow: hidden;

  &&:not([hidden]) {
    display: flex;
  }
`;
const DialogLayout = styled(Flex) `
  flex: 1;
  min-height: 0;
`;
const DialogHeader = styled(Card) `
  position: relative;

  &:after {
    content: '';
    display: block;
    position: absolute;
    left: 0;
    right: 0;
    bottom: -1px;
    border-bottom: 1px solid var(--card-hairline-soft-color);
  }
`;
const DialogContent = styled(Box) `
  overflow: auto;
  outline: none;
`;
const DialogFooter = styled(Box) `
  border-top: 1px solid var(--card-hairline-soft-color);
`;
const DialogCard = forwardRef(({ cardRadius, cardShadow, children, footer, header, id, onClose, scheme, width, }, ref) => {
    const [rootElement, setRootElement] = useState(null);
    const contentRef = useRef(null);
    const { isTopLayer } = useLayer();
    const labelId = `${id}_label`;
    useEffect(() => {
        // On mount: focus the first interactive element in the contents
        if (contentRef.current) {
            focusFirstDescendant(contentRef.current);
        }
    }, []);
    useGlobalKeyDown(useCallback((event) => {
        if (!isTopLayer)
            return;
        if (event.key === 'Escape') {
            event.preventDefault();
            event.stopPropagation();
            if (onClose)
                onClose();
        }
    }, [isTopLayer, onClose]));
    useClickOutside(useCallback(() => {
        if (!isTopLayer)
            return;
        if (onClose)
            onClose();
    }, [isTopLayer, onClose]), [rootElement]);
    const setRef = useCallback((el) => {
        setRootElement(el);
        if (typeof ref === 'function')
            ref(el);
        else if (ref)
            ref.current = el;
    }, [ref]);
    return (React.createElement(DialogContainer, { width: width },
        React.createElement(DialogCardRoot, { radius: cardRadius, ref: setRef, scheme: scheme, shadow: cardShadow },
            React.createElement(DialogLayout, { direction: "column" },
                React.createElement(DialogHeader, null,
                    React.createElement(Flex, null,
                        React.createElement(Box, { flex: 1, padding: 4 }, header && (React.createElement(Text, { id: labelId, weight: "semibold" }, header))),
                        React.createElement(Box, { padding: 2 },
                            React.createElement(Button, { "aria-label": "Close dialog", icon: "close", mode: "bleed", onClick: onClose, padding: 3 })))),
                React.createElement(DialogContent, { flex: 1, ref: contentRef, tabIndex: -1 }, children),
                footer && React.createElement(DialogFooter, null, footer)))));
});
DialogCard.displayName = 'DialogCard';
export const Dialog = forwardRef((props, ref) => {
    const { cardRadius = 2, cardShadow = 4, children, footer, header, id, onClose, scheme, width = 0 } = props, restProps = __rest(props, ["cardRadius", "cardShadow", "children", "footer", "header", "id", "onClose", "scheme", "width"]);
    const preDivRef = useRef(null);
    const postDivRef = useRef(null);
    const cardRef = useRef(null);
    const handleFocus = useCallback((event) => {
        const target = event.target;
        const cardElement = cardRef.current;
        if (!cardElement) {
            return;
        }
        if (target === preDivRef.current) {
            focusLastDescendant(cardElement);
            return;
        }
        if (target === postDivRef.current) {
            focusFirstDescendant(cardElement);
            return;
        }
    }, []);
    const labelId = `${id}_label`;
    return (React.createElement(Portal, null,
        React.createElement(Root, Object.assign({}, restProps, { "aria-labelledby": labelId, "aria-modal": true, id: id, onFocus: handleFocus, ref: ref, role: "dialog" }),
            React.createElement("div", { ref: preDivRef, tabIndex: 0 }),
            React.createElement(DialogCard, { cardRadius: cardRadius, cardShadow: cardShadow, footer: footer, header: header, id: id, onClose: onClose, ref: cardRef, scheme: scheme, width: width }, children),
            React.createElement("div", { ref: postDivRef, tabIndex: 0 }))));
});
Dialog.displayName = 'Dialog';
