// @sanity/ui/dist/cjs/atoms/popover/popover.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Popover = void 0;
var react_1 = __importStar(require("react"));
var react_popper_1 = require("react-popper");
var utils_1 = require("../../utils");
var card_1 = require("../card");
var arrow_1 = require("./arrow");
exports.Popover = react_1.forwardRef(function (props, ref) {
    var boundaryElementContext = utils_1.useBoundaryElement();
    var _a = props.boundaryElement, boundaryElementProp = _a === void 0 ? boundaryElementContext : _a, child = props.children, content = props.content, disabled = props.disabled, open = props.open, padding = props.padding, placementProp = props.placement, _b = props.portal, portalProp = _b === void 0 ? true : _b, _c = props.radius, radius = _c === void 0 ? 3 : _c, referenceElementProp = props.referenceElement, _d = props.shadow, shadow = _d === void 0 ? 3 : _d, scheme = props.scheme, _e = props.style, style = _e === void 0 ? {} : _e, restProps = __rest(props, ["boundaryElement", "children", "content", "disabled", "open", "padding", "placement", "portal", "radius", "referenceElement", "shadow", "scheme", "style"]);
    var placement = typeof placementProp === 'string' ? placementProp : 'bottom';
    var portal = utils_1.usePortal();
    var boundaryElement = boundaryElementProp || portal.boundaryElement;
    var _f = react_1.useState(null), referenceElement = _f[0], setReferenceElement = _f[1];
    var _g = react_1.useState(null), popperElement = _g[0], setPopperElement = _g[1];
    var _h = react_1.useState(null), arrowElement = _h[0], setArrowElement = _h[1];
    var popperReferenceElement = referenceElementProp || referenceElement;
    var popper = react_popper_1.usePopper(popperReferenceElement, popperElement, {
        placement: placement,
        modifiers: [
            {
                name: 'arrow',
                options: {
                    element: arrowElement,
                    padding: 4,
                },
            },
            {
                name: 'preventOverflow',
                options: {
                    altAxis: true,
                    boundary: boundaryElement || undefined,
                    padding: 8,
                },
            },
            {
                name: 'offset',
                options: {
                    offset: [0, 4],
                },
            },
        ],
    });
    var attributes = popper.attributes, forceUpdate = popper.forceUpdate, styles = popper.styles;
    react_1.useEffect(function () {
        if (forceUpdate)
            forceUpdate();
    }, [forceUpdate, content, popperReferenceElement]);
    if (disabled) {
        return child || react_1.default.createElement(react_1.default.Fragment, null);
    }
    var setRef = function (el) {
        var childRef = child.ref;
        setReferenceElement(el);
        if (typeof childRef === 'function') {
            childRef(el);
        }
        else if (childRef) {
            childRef.current = el;
        }
    };
    var setRootRef = function (el) {
        setPopperElement(el);
        if (typeof ref === 'function')
            ref(el);
        else if (ref)
            ref.current = el;
    };
    var node = (react_1.default.createElement(utils_1.Layer, __assign({ "data-ui": "Popover" }, restProps, { ref: setRootRef, style: __assign(__assign(__assign({}, style), styles.popper), { pointerEvents: 'all' }) }, attributes.popper),
        react_1.default.createElement(card_1.Card, { "data-ui": "PopoverCard", padding: padding, radius: radius, scheme: scheme, shadow: shadow },
            react_1.default.createElement(arrow_1.PopoverArrow, { ref: setArrowElement, tone: "default", style: styles.arrow }),
            content)));
    return (react_1.default.createElement(react_1.default.Fragment, null,
        child && !referenceElementProp ? react_1.cloneElement(child, { ref: setRef }) : child || react_1.default.createElement(react_1.default.Fragment, null),
        open && (react_1.default.createElement(react_1.default.Fragment, null,
            portalProp && react_1.default.createElement(utils_1.Portal, null, node),
            !portalProp && node))));
});
exports.Popover.displayName = 'Popover';
